### DISCUSSION TITLE
Final FTD syntax

### DISCUSSION THREAD

**Discussion by Arpita-Jaiswal** [fastn-stack/fastn#429] [General]
FTD language has five things:
- Import
- Record
- Variable
- Component
- Comments

# Import

The `import` keyword is used to make code in one module available to another. The syntax for `import` is

```ftd
-- 𝗶𝗺𝗽𝗼𝗿𝘁: 𝗳𝗼𝗼
```

It also accepts alias separated by the `as` keyword.
Consider this

```ftd
  -- import: foo 𝗮𝘀 𝗳
```
Here `f` is an alias for `foo`. i.e. We can use `f` instead of `foo` in the current module.

# Record

The `record` is a type that is composed of other types. We can define a `record` in the following way:

```ftd
-- record person:
caption name:
integer age:
string company:
```

In the above example, we have defined the record ` person` with some other fields. Record fields can contain default values too.
Consider this

```ftd
-- record person:
caption name:
integer age:
string company: FifthTry
```

Here `FifthTry` is the default value for the field `company`. If the variable is initialized without the field company, the default value would become the value for the field 

## Record Methods

```ftd
-- record person:
string first-name:
string last-name:

-- string person.full-name:

first-name + last-name

-- end: full-name

-- end: person
```

```ftd
-- record person:
string first-name:
string last-name:

-- string list person.location:

-- string: Varanasi
-- string: Bengaluru

-- end: person.location

-- string list person.full-name():

[first-name + last-name]

-- end: person.full-name

-- end: person
```

## Record Views

```ftd
-- record person:
string first-name:
string last-name:

-- string person.full-name():

first-name + last-name

-- end: person.full-name

-- component person.display:

-- ftd.text: 

{ person.full-name() } is the full name.

-- end: display

-- end: person
```

Invoking record view

```ftd
-- person arpita:
first-name: Arpita
last-name: Jaiswal

-- arpita.display:

```

# Variable

The `variable` helps to store the value. In FTD, variables are associated with a specific data type. Variable can be defined in the following way:

```ftd
;; Case #1
-- string name: Arpita

;; Case #2
-- person arpita:
name: Arpita
age: 10
```

In case 1, the `name` is the variable name for this variable which is of type `string` with value as `Arpita`.
In case 2, the `arpita` variable is of type `person` having fields with values as `name: Arpita` and `age: 10`. Note variable `arpita` has not given any value for field `company`. As discussed above, `company` would take the default value `FifthTry` for this variable.

### Immutable

By default variables in FTD are immutable i.e. read-only. In other words, a variable value cannot be changed once a value is bound to a variable name.

Let us understand this with an example.

```ftd
-- string name: Arpita

-- $name: Ayushi
```

The above code throws an error because you cannot assign values twice to an immutable variable.

# Mutable

The variable defined as mutable can change its value multiple times. Prefix the variable name`$` to make a variable mutable.

Consider the below example

```ftd
-- string $name: Arpita

-- $name: Ayushi
```

The above code compiles as the `name` variable is mutable.

### Variable Definition

The variable can be defined with the following blueprint.

```ftd
;; case #1: variable_value in caption
-- <variable_type> <variable_name>: <variable_value>


;; case #2: variable_value in header
-- <variable_type> <variable_name>:
<variable_value> ;; as header (key value pair)

;; case #3 variable_value in body
-- <variable_type> <variable_name>:

<variable_value>
```

Consider the two examples above for mutable and immutable definitions of variables:

```ftd
-- person arpita:
name: Arpita
age: 10

-- string $name: Arpita
```

### Variable Invocation/Updation

The variable defined as mutable can be updated/invocated. Consider the code blueprint below

```ftd
-- $<variable_name>: <new variable value>
```

Let's consider an example:

```ftd
-- $name: Ayushi
```

# Component

Components are independent and reusable bits of code. They accept the arguments that give some customizability.
When creating an FTD component, it should be started with the `component` keyword, followed by the component name.
Then it defines its bunch of arguments in `headers` followed by the component type or Component Tree.

Consider the following example:

```ftd
-- component print: 
caption name:

-- ftd.text: $name

-- end: print
```

In the above example, `print` is the component name of type `ftd.text`. It accepts the `name` argument which is of type `caption`(an alias for string)

We can invoke this component in the following way

```ftd
-- print: 
name: Arpita
```

In the above code, we pass the value `Arpita` to the component's argument of type `caption`. This is called the `property` of the component. i.e During the invocation of the component, the component's `argument` becomes the invocation `property`. Hence the `name` is the property here.

## Simple component type component

The simple component is the component that doesn't accept children during its invocation. Like in the case of the above example, `print` doesn't have any children.

## Open container type component

The open container component accepts the children during invocation. Consider the example below

```ftd
-- component show:
string name:
ftd.ui list children:

  -- ftd.row:
  children: $children
  
  -- ftd.text: $name
  
  -- end: ftd.row

-- end: show
```

Here `ftd.ui` takes any component as value to the property `children`. Also `children` (in argument) is a reserved keyword which refers to the children passed in the subsection during invocation. The components as the value passed to the `children` property become the children of `ftd.row`. This is ensured by the `children` property of `ftd.row`.
Let's see how to invoke the above component.

```ftd
-- show:
name: Arpita

-- ftd.column:
  -- ftd.text: Hello
  -- ftd.text: World
-- end: ftd.column

-- ftd.text: ✌️

-- end: show
```

In the above example, the `ftd.column` and `ftd.text` becomes the children of `ftd.row` defined in `show`.

### Passing components in headers

We can have multiple arguments which are of `ftd.ui` or `ftd.ui list` type. This gives us the ability to append components in multiple places. Consider the example below

```ftd
-- component page:
ftd.ui list children:
ftd.ui list headers:

-- ftd.column:
  ;; Way# 1
  -- header:
  children: $headers

  ;; Way# 2
  -- $obj:
  $loop$: $headers as $obj
  
  -- ftd.column:
  children: $children
-- end: ftd.column

-- end: page
```

Let's see its invocation:

```ftd
-- page: 

-- page.headers:
  -- ftd.text: This is the value of the `headers` property
-- end: page.headers

--  ftd.text: Hello
-- ftd.text: world

-- end: page
```
 
In the above code, the `-- ftd.text: This is the value of `headers` property is passed as property of `headers` and 
```
--  ftd.text: Hello
-- ftd.text: world
```
becomes the value of `children` property.

### More about component arguments

The component can define arguments with the default value. Consider the following code:

```ftd
-- component print:
string name: Hello World

-- ftd.text: $name

-- end: print


;; The output for this is `Hello World`
-- print:


;; The output for this is `Hello There`
-- print:
name: Hello There
```

#### Immutable Arguments

By default arguments are immutable. Consider this example.

```ftd
-- component print:
string name: Hello World

-- ftd.text: $name

-- end: print
```

Here `name` is immutable.

### Mutable Arguments

We can make an argument mutable by adding the prefix `$` to the argument name. Consider the code below.

```ftd
-- component print:
string $name: Hello World

-- ftd.text: $name

-- ftd.text: Click Here
$on-click$: $name = Hello again

-- end: print
```

In the above code, the `name` becomes mutable. A mutable argument can accept the mutable variable and can change the variable value. `$on-click$: $name = Hello again` changes the value of argument `name` on click event.

```ftd
;; Case #0
-- print:

;; Case #1
-- print:
$name: Arpita


;; Case #2 
-- string $msg: The message from msg.

-- print:
$name: $msg

-- ftd.text: $msg
```

In case 1, the property `name` is passed with the value `Arpita`. This overrides the default value `Hello World` of the argument `name`.

In case 2, the property `name` is passed with the mutable reference of variable `msg`. The `$on-click$` event would change the value of `msg` too. So both inside component `print` definition (`-- ftd.text: $name`) and outside it, in invocation, `-- ftd.text: $msg`. The value would show the new value set by the event (`Hello again`).

Output before `$on-click$` event
```ftd
;; Output from case #0
Hello World

;; Output from case #1
Arpita

;; Output from case #2 
The message from msg   ;; by `print` invocation

The message from msg   ;; by `-- ftd.text: $msg`
```

Output after `$on-click$` event
```ftd
;; Output from case #0
Hello again

;; Output from case #1
Hello again

;; Output from case #2 
Hello again  ;; by `print` invocation

Hello again  ;; by `-- ftd.text: $msg`
```

# Comments
There are two ways to comment in `ftd`.

## Commenting whole `Section`

With `/` in front of `-- ftd.row` will comment out the whole section till `-- end: ftd.row`.
 
```ftd
/-- ftd.row:

-- ftd.text: Hello World with Some Comments

-- end: ftd.row
```

## Inline Comments

The `;;` will comment out the whole line.

```ftd

;; -- ftd.text: Hello World

;; Commented line

```

# Create the inline list
# Update mutable list
# Record containing list
# Nested list (record containing list containing record)
# Leaf (Non-container) and container component
# Container invocation with an attribute being another component/ caption being another component
# attribute being a list of component
# attribute being Inline record /Inline list
# conditional attributes
# Function signature

```ftd
-- integer sum():
integer a:
integer b:
``` 

By default, the function without a body is a foreign function.

```ftd
-- integer c: { sum(a: 2, b: 3) }
```

```ftd
-- integer list_sum():
integer list l:

-- integer c: { list_sum(l: [2, 3])}
```

```ftd
-- record person:
string name:
integer age:

-- integer list_sum():
person list l:

-- person amitu:
name: AmitU
age: 19

-- integer c: { list_sum(
      l: [
        person {
            name: "Arpita",
            age: 10
        },
        amitu
      ]
    )}
    
-- integer c: 

{ list_sum(
    l: [
      person {
          name: "Arpita",
          age: 10
      },
      amitu
    ]
)}
```

## Event Handler functions

```ftd
-- void increment():
integer by: 1
integer $c:

$c = $c + by

-- integer $number: 1

-- ftd.text:  $number
$on-click$: { increment(by: 2, $c: number) }
```

**Comment by AbrarNitk** [comment]
Awesome!! @Arpita-Jaiswal

**Comment by Arpita-Jaiswal** [comment]
How to comment out multiple sections and multiple lines??
